File: pat/a/1028.rb
#!/usr/bin/env ruby
# TLE on the last case

class Student
  attr_accessor :num, :name, :score, :numi
  def initialize(line)
    words = line.split ' '
    @num = words[0]
    @numi = @num.to_i
    @name = words[1]
    @score = words[2].to_i
  end

  def <=>(other)
    @numi <=> other.numi
  end

  def to_s
    "#{@num} #{@name} #{@score}"
  end
end

def compare(a, b)
  res = yield(a, b)
  if res == 0
    a <=> b
  else
    res
  end
end

if __FILE__ == $0
  lines = STDIN.readlines
  field = lines[0].split(' ')[1].to_i
  lines = lines[1,lines.size - 1]
  students = lines.map { |s| Student.new s }
  if field == 1
    students.sort!
  elsif field == 2
    students.sort! do |x, y|
      compare(x, y) do |a, b|
        a.name <=> b.name
      end
    end
  else
    students.sort! do |x, y|
      compare(x, y) do |a, b|
        a.score <=> b.score
      end
    end
  end
  puts students
end

File: pat/a/1028.hs
-- MLE on the last case
import Control.Monad
import qualified Data.List as List

data Student = Student { num :: String,
                         name :: String,
                         score :: Int  }

instance Eq Student where
    (Student num1 _ _) == (Student num2 _ _) = num1 == num2

instance Ord Student where
    compare (Student num1 _ _) (Student num2 _ _) = compare num1 num2 

instance Show Student where
    show (Student num name score) = num ++ " " ++ name ++ " " ++ show score

compareByName :: Student -> Student -> Ordering
compareByName s1@(Student _ name1 _) s2@(Student _ name2 _)
    = if nameOrd == EQ then compare s1 s2 else nameOrd
      where nameOrd = compare name1 name2

compareByScore :: Student -> Student -> Ordering
compareByScore s1@(Student _ _ score1) s2@(Student _ _ score2)
    = if scoreOrd == EQ then compare s1 s2 else scoreOrd
      where scoreOrd = compare score1 score2 

main = do fstLine <- getLine
          let lwords = words fstLine  
              (totalStudents, field) = (read (lwords !! 0) :: Int, read (lwords !! 1) :: Int) 
          students <- forM [i | i <- [1..totalStudents]] (\_ -> do
            line <- getLine
            let sWords = words line
            return $ Student (sWords !! 0) (sWords !! 1) (read $ sWords !! 2))
          let comparer = [compare, compareByName, compareByScore]
          mapM_ print (List.sortBy (comparer !! (field-1)) students)
          
